Add a blog post detailing how to solve intake-esm aggregation error (#56)
NCAR/esds: posts/intake_cmip6_debug.md

```python
# Set the path for the catalog file
catalog_file = '/glade/collections/cmip/catalog/intake-esm-datastore/catalogs/glade-cmip6.json'

# Open the catalog using intake
col = intake.open_esm_datastore(catalog_file)
```

Now, we set which experiment to use. In this case, we are interested in the Ocean Model Intercomparison Project (OMIP) since we are looking at ocean data, setting

```python
experiment_id = ['omip1']
```

In terms of the variables of interest, we are interested in Sea Water Silinity (`so`) and Sea Water Potential Temperature (`thetao`), so we set

```python
variable_id = ['so', 'thetao']
```

The last variable to change is the table_id which corresponds to the temporal frequency. Since we are interested in monthly data, we set

```python
table_id = 'Omon'
```

Putting all that together, we setup the catalog and call `to_dataset_dict` which assembles a dictionary of datasets

```python
cat = col.search(
    experiment_id=['omip1'],
    variable_id=['thetao', 'so'],
    table_id='Omon'
)
dset_dict = cat.to_dataset_dict()
```

> But this returns an error of:

Reviewer: As of the time of this post, this is the error returned. But note: this is a problem with the underlying data available on disk—so the same code may cease to return this particular error at some future point.

I suggest making this clear.

```
AggregationError:
        Failed to merge multiple datasets in group with key=OMIP.CNRM-CERFACS.CNRM-CM6-1.omip1.Omon.gn into a single xarray Dataset as variables.

        *** Arguments passed to xarray.merge() ***:

        - objs: a list of 2 datasets
        - kwargs: {}

        ********************************************
```

## Determining the Cause of the Error

At this point, let's go back and inspect the subset catalog we have, calling the `catalog.df`

```python
cat.df
```

Notice how in the error, the source_id which triggered the error is `CNRM-CM6-1`, so let's subset for that and investigate the problem...

```python

> cat.df[cat.df.source_id == 'CNRM-CM6-1']

Reviewer: ```suggestion
cat.df.loc[cat.df.source_id == 'CNRM-CM6-1']
```

Reviewer 2: You could bypass the pandas indexer by using the problematic key

```python
cat['OMIP.CNRM-CERFACS.CNRM-CM6-1.omip1.Omon.gn']
```

This returns an object with the exact files that belong to the problematic key... This object has a corresponding `.df`

```

> 

Reviewer: Are you planning to include the DataFrame ouput?

Author: As of now, the site still only supports markdown.. working on a PR to add notebook functionality since having the ability to view dataframes inline would helpful/along with making it easier write the markdown + code

Here, we focus on the `time_range` column - noticing that the last time step for `so` is `194912` while the last timestep for datasets with `thetao` is `199912`, with the dates formatted `YYYYMM`.

> When intake attempts to concatentate these two datasets, with different times, it struggles and returns the error that it cannot merge them together since the length of the time axis differs.

Reviewer: ```suggestion
When `intake` attempts to concatenate these two datasets, it struggles and return an error, since it cannot merge datasets where the length of the time axis differs.
```

> 

Reviewer: Here is where I think is important to show—or at least mention—the `aggegrate=False` solution.

Author: I mention it below - would that be okay?

> ## Applying our Solution

Reviewer: It's not so much a "solution" as a "workaround." Fundamentally, there is data missing! 

I would suggest presenting "reading variables separately" as another approach to debugging. Doing this "works" in that a dataset is returned...but still: there's missing data!

To fix this problem, If you would like to still work with the data, you will need to read in datasets using separate queries as shown below

```python
# Search and read in dataset for so
cat_so = col.search(
    experiment_id=['omip1'],
    variable_id=['so'],
    table_id='Omon'
)
dset_dict_so = cat_so.to_dataset_dict()

# Search and read in dataset for thetao
cat_thetao = col.search(
    experiment_id=['omip1'],
    variable_id=['thetao'],
    table_id='Omon'
)
dset_dict_thetao = cat_thetao.to_dataset_dict()
```

> Another option would have been to specify within .to_dataset_dict() to not aggregate the datasets, using the following syntax .to_dataset_dict(aggregate=False), but it can be difficult using this functionality with such a large number of ensemble members.

Reviewer: Fine to leave this here, I guess. 

```suggestion
Another option is to turn off aggregation within `to_dataset_dict()`, using the following syntax 
`dsets = cat.to_dataset_dict(aggregate=False)`
This will return a dataset for **every** file in the archive and the keys in `dset_dict` will be constructed using all the fields in the catalog. Since `aggregate=False` will yield a large number of individual datasets, it might be overwhelming and hard to determine the problem.
```
